{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5f219093",
   "metadata": {},
   "source": [
    "# In this notebook we divide our data and assigning each part to a Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c7ae3b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "from ipywidgets import interact\n",
    "import ipywidgets as widgets\n",
    "import plotly.graph_objects as go\n",
    "from imblearn.over_sampling import SMOTE\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix, ConfusionMatrixDisplay\n",
    "from sklearn.decomposition import PCA\n",
    "from mpl_toolkits.mplot3d import Axes3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "598171c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>geographical_location</th>\n",
       "      <th>family_history</th>\n",
       "      <th>smoking_habits</th>\n",
       "      <th>alcohol_consumption</th>\n",
       "      <th>helicobacter_pylori_infection</th>\n",
       "      <th>dietary_habits</th>\n",
       "      <th>existing_conditions</th>\n",
       "      <th>endoscopic_images</th>\n",
       "      <th>biopsy_results</th>\n",
       "      <th>ct_scan</th>\n",
       "      <th>mature_mirna_acc</th>\n",
       "      <th>mature_mirna_id</th>\n",
       "      <th>target_symbol</th>\n",
       "      <th>target_entrez</th>\n",
       "      <th>target_ensembl</th>\n",
       "      <th>diana_microt</th>\n",
       "      <th>elmmo</th>\n",
       "      <th>microcosm</th>\n",
       "      <th>miranda</th>\n",
       "      <th>mirdb</th>\n",
       "      <th>pictar</th>\n",
       "      <th>pita</th>\n",
       "      <th>targetscan</th>\n",
       "      <th>predicted.sum</th>\n",
       "      <th>all.sum</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>Other</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Low_Salt</td>\n",
       "      <td>Chronic Gastritis</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>5034</td>\n",
       "      <td>1946626</td>\n",
       "      <td>0.3813</td>\n",
       "      <td>0.1870</td>\n",
       "      <td>0.7864</td>\n",
       "      <td>0.2048</td>\n",
       "      <td>0.5619</td>\n",
       "      <td>0.4382</td>\n",
       "      <td>0.2836</td>\n",
       "      <td>0.9282</td>\n",
       "      <td>4.3243</td>\n",
       "      <td>7.6668</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>86</td>\n",
       "      <td>Female</td>\n",
       "      <td>Ethnicity_B</td>\n",
       "      <td>California</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Diabetes</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR234_2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>6901</td>\n",
       "      <td>1503178</td>\n",
       "      <td>0.9586</td>\n",
       "      <td>0.4933</td>\n",
       "      <td>0.9640</td>\n",
       "      <td>0.4980</td>\n",
       "      <td>0.9856</td>\n",
       "      <td>0.1446</td>\n",
       "      <td>0.3754</td>\n",
       "      <td>0.1036</td>\n",
       "      <td>7.9677</td>\n",
       "      <td>1.4833</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>68</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>California</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>MIR345</td>\n",
       "      <td>MIR345_3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>4014</td>\n",
       "      <td>1787909</td>\n",
       "      <td>0.2695</td>\n",
       "      <td>0.5736</td>\n",
       "      <td>0.6669</td>\n",
       "      <td>0.5404</td>\n",
       "      <td>0.9059</td>\n",
       "      <td>0.8273</td>\n",
       "      <td>0.3509</td>\n",
       "      <td>0.1669</td>\n",
       "      <td>3.7487</td>\n",
       "      <td>3.0468</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57</td>\n",
       "      <td>Female</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>Other</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Chronic Gastritis</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>7351</td>\n",
       "      <td>1310766</td>\n",
       "      <td>0.3723</td>\n",
       "      <td>0.2614</td>\n",
       "      <td>0.9495</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.4299</td>\n",
       "      <td>0.9352</td>\n",
       "      <td>0.7947</td>\n",
       "      <td>0.8670</td>\n",
       "      <td>5.4783</td>\n",
       "      <td>8.8113</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>California</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Diabetes</td>\n",
       "      <td>Abnormal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>MIR345</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>CDH1</td>\n",
       "      <td>7982</td>\n",
       "      <td>1277058</td>\n",
       "      <td>0.9747</td>\n",
       "      <td>0.7545</td>\n",
       "      <td>0.2632</td>\n",
       "      <td>0.8768</td>\n",
       "      <td>0.6508</td>\n",
       "      <td>0.3377</td>\n",
       "      <td>0.4275</td>\n",
       "      <td>0.9158</td>\n",
       "      <td>1.8092</td>\n",
       "      <td>0.3946</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age  gender    ethnicity geographical_location  family_history  \\\n",
       "0   43    Male  Ethnicity_A                 Other               1   \n",
       "1   86  Female  Ethnicity_B            California               1   \n",
       "2   68    Male  Ethnicity_A            California               0   \n",
       "3   57  Female  Ethnicity_A                 Other               0   \n",
       "4   33    Male  Ethnicity_A            California               0   \n",
       "\n",
       "   smoking_habits  alcohol_consumption  helicobacter_pylori_infection  \\\n",
       "0               0                    0                              0   \n",
       "1               0                    0                              1   \n",
       "2               1                    1                              0   \n",
       "3               0                    0                              1   \n",
       "4               1                    1                              0   \n",
       "\n",
       "  dietary_habits existing_conditions endoscopic_images biopsy_results  \\\n",
       "0       Low_Salt   Chronic Gastritis            Normal       Negative   \n",
       "1      High_Salt            Diabetes            Normal       Negative   \n",
       "2      High_Salt                 NaN            Normal       Negative   \n",
       "3      High_Salt   Chronic Gastritis            Normal       Negative   \n",
       "4      High_Salt            Diabetes          Abnormal       Negative   \n",
       "\n",
       "    ct_scan mature_mirna_acc mature_mirna_id target_symbol  target_entrez  \\\n",
       "0  Negative           MIR123        MIR123_1          TP53           5034   \n",
       "1  Negative           MIR123        MIR234_2          TP53           6901   \n",
       "2  Negative           MIR345        MIR345_3          KRAS           4014   \n",
       "3  Negative           MIR123        MIR123_1          KRAS           7351   \n",
       "4  Negative           MIR345        MIR123_1          CDH1           7982   \n",
       "\n",
       "   target_ensembl  diana_microt   elmmo  microcosm  miranda   mirdb  pictar  \\\n",
       "0         1946626        0.3813  0.1870     0.7864   0.2048  0.5619  0.4382   \n",
       "1         1503178        0.9586  0.4933     0.9640   0.4980  0.9856  0.1446   \n",
       "2         1787909        0.2695  0.5736     0.6669   0.5404  0.9059  0.8273   \n",
       "3         1310766        0.3723  0.2614     0.9495   0.1342  0.4299  0.9352   \n",
       "4         1277058        0.9747  0.7545     0.2632   0.8768  0.6508  0.3377   \n",
       "\n",
       "     pita  targetscan  predicted.sum  all.sum  label  \n",
       "0  0.2836      0.9282         4.3243   7.6668      0  \n",
       "1  0.3754      0.1036         7.9677   1.4833      0  \n",
       "2  0.3509      0.1669         3.7487   3.0468      0  \n",
       "3  0.7947      0.8670         5.4783   8.8113      0  \n",
       "4  0.4275      0.9158         1.8092   0.3946      0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_data = pd.read_csv('../datasets/MainDataset/gastric_cancer_detection_dataset.csv')\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.precision', 4)\n",
    "\n",
    "gc_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95f5da21",
   "metadata": {},
   "source": [
    "## Understanding the Dataset (Full Explanation)\n",
    "\n",
    "In this step, we explain each feature in the dataset in a way that is accessible even to those without a background in medicine, biology, or genetics.\n",
    "\n",
    "### ðŸ”¹ Demographic & Lifestyle Information\n",
    "\n",
    "These features capture general information about the patient and their habits, which may influence the risk of developing gastric cancer.\n",
    "\n",
    "- **age** *(Numerical)*: The age of the patient at the time of diagnosis or evaluation.\n",
    "- **gender** *(Categorical: Male/Female)*: The biological sex of the patient.\n",
    "- **ethnicity** *(Categorical)*: The ethnic background of the patient (e.g., Asian, Caucasian).\n",
    "- **geographical_location** *(Categorical)*: The country or region where the data was collected. Environmental or regional factors can influence cancer rates.\n",
    "- **family_history** *(Binary: Yes/No)*: Whether the patient has a family history of cancer, which may increase their risk.\n",
    "- **smoking_habits** *(Binary: Yes/No)*: Indicates if the patient is a smoker, a known risk factor for gastric and many other cancers.\n",
    "- **alcohol_consumption** *(Binary: Yes/No)*: Indicates whether the patient consumes alcohol, which can also be a risk factor.\n",
    "- **dietary_habits** *(Categorical)*: Describes the patientâ€™s eating patternsâ€”such as diets high in saltâ€”which can affect cancer risk.\n",
    "\n",
    "### ðŸ”¹ Medical History & Diagnostic Results\n",
    "\n",
    "These features involve the patient's medical history and results from medical tests that are commonly used to detect cancer.\n",
    "\n",
    "- **existing_conditions** *(Categorical)*: Any pre-existing medical conditions like diabetes, gastritis, or hypertension, which may be related to overall health and cancer risk.\n",
    "- **helicobacter_pylori_infection** *(Binary: Yes/No)*: Indicates whether the patient is infected with *Helicobacter pylori*, a bacteria strongly linked to stomach cancer.\n",
    "- **endoscopic_images** *(Categorical: Normal/Abnormal)*: The result of an endoscopy, which uses a camera to look inside the stomach. An abnormal result may show signs of tumors or ulcers.\n",
    "- **biopsy_results** *(Categorical: Positive/Negative)*: Whether a tissue sample taken from the stomach tested positive (cancerous) or negative (non-cancerous).\n",
    "- **ct_scan** *(Categorical: Normal/Abnormal)*: Results from a CT (computed tomography) scan. Abnormal results may indicate the presence of a tumor.\n",
    "\n",
    "### ðŸ”¹ Genetic & Molecular Data (miRNA and Target Genes)\n",
    "\n",
    "These columns relate to *microRNAs (miRNAs)*, which are small non-coding RNA molecules that regulate gene expression. Changes in miRNA behavior can be linked to cancer. Each miRNA can bind to target genes and affect how much of the protein is produced.\n",
    "\n",
    "- **mature_mirna_acc** *(Categorical)*: A unique identifier (accession number) for the mature miRNA.\n",
    "- **mature_mirna_id** *(Categorical)*: The name of the mature miRNA, such as *hsa-miR-21*, where \"hsa\" refers to humans.\n",
    "- **target_symbol** *(Categorical)*: The shorthand symbol for a gene that the miRNA might regulate.\n",
    "- **target_entrez** *(Categorical)*: A unique numerical ID for the gene in the NCBI Entrez database.\n",
    "- **target_ensembl** *(Categorical)*: A unique identifier for the gene in the Ensembl database, another widely used genomic database.\n",
    "\n",
    "### ðŸ”¹ miRNAâ€“Gene Interaction Prediction Scores\n",
    "\n",
    "These columns come from bioinformatics tools that predict whether a miRNA can interact with a specific gene. Each tool uses different algorithms or biological data to make these predictions. Higher scores generally indicate stronger or more confident interactions.\n",
    "\n",
    "- **diana_microt** *(Numerical)*: Score from the DIANA-microT algorithm.\n",
    "- **elmmo** *(Numerical)*: Confidence score from the ElMMo tool.\n",
    "- **microcosm** *(Numerical)*: Score from the Microcosm prediction database.\n",
    "- **miranda** *(Numerical)*: Score from the MiRanda alignment-based algorithm.\n",
    "- **mirdb** *(Numerical)*: Confidence score from the miRDB database.\n",
    "- **pictar** *(Numerical)*: Predicted interaction score from the PicTar algorithm.\n",
    "- **pita** *(Numerical)*: Score based on how accessible a gene is to a miRNA (from the PITA algorithm).\n",
    "- **targetscan** *(Numerical)*: Score from TargetScan, one of the most widely used miRNA target prediction tools.\n",
    "\n",
    "- **predicted.sum** *(Numerical)*: Sum of scores across prediction algorithms. Indicates overall prediction confidence from computational tools.\n",
    "- **all.sum** *(Numerical)*: May include additional scoring beyond prediction tools, such as experimentally validated data.\n",
    "\n",
    "### ðŸŽ¯ Target Variable\n",
    "\n",
    "- **label** *(Binary: 1 = Cancer, 0 = No Cancer)*: This is the **output** or **label** column we aim to predict using machine learning. It tells whether the patient has been diagnosed with gastric cancer.\n",
    "\n",
    "---\n",
    "\n",
    "### ðŸ§  Summary\n",
    "This dataset combines:\n",
    "- Basic demographic and lifestyle risk factors\n",
    "- Clinical and diagnostic test results\n",
    "- Advanced genomic data related to miRNAs and their gene targets\n",
    "\n",
    "Together, this rich set of features allows for a powerful machine learning model that can potentially detect gastric cancer based on diverse types of input data.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f306c0fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['age', 'gender', 'ethnicity', 'geographical_location', 'family_history',\n",
       "       'smoking_habits', 'alcohol_consumption',\n",
       "       'helicobacter_pylori_infection', 'dietary_habits',\n",
       "       'existing_conditions', 'endoscopic_images', 'biopsy_results', 'ct_scan',\n",
       "       'mature_mirna_acc', 'mature_mirna_id', 'target_symbol', 'target_entrez',\n",
       "       'target_ensembl', 'diana_microt', 'elmmo', 'microcosm', 'miranda',\n",
       "       'mirdb', 'pictar', 'pita', 'targetscan', 'predicted.sum', 'all.sum',\n",
       "       'label'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gc_data.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1302332e",
   "metadata": {},
   "source": [
    "# ðŸ“Š Feature Division by Model\n",
    "\n",
    "We divide the dataset features into three models based on the stage of detection and the nature of the data.\n",
    "\n",
    "---\n",
    "\n",
    "## ðŸ”¹ Model 1: Early Risk Screening\n",
    "**Goal:** Predict early risk based on lifestyle, demographic, and infection data.\n",
    "\n",
    "**Used Features:**\n",
    "- `Age`\n",
    "- `Gender`\n",
    "- `Ethnicity`\n",
    "- `Geographical Location`\n",
    "- `Family History`\n",
    "- `Smoking Habits`\n",
    "- `Alcohol Consumption`\n",
    "- `Helicobacter Pylori Infection`\n",
    "- `Dietary Habits`\n",
    "- `Existing Conditions`\n",
    "\n",
    "---\n",
    "\n",
    "## ðŸ”¹ Model 2: Genetic Risk Model\n",
    "**Goal:** Use genetic/molecular features to assess genetic predisposition among \"safe\" users from Model 1.\n",
    "\n",
    "**Used Features:**\n",
    "- `Mature miRNA Accession`\n",
    "- `Mature miRNA ID`\n",
    "- `Target Symbol`\n",
    "- `Target Entrez`\n",
    "- `Target Ensembl`\n",
    "- `DIANA microT Score`\n",
    "- `ElMMo Score`\n",
    "- `Microcosm Score`\n",
    "- `MiRanda Score`\n",
    "- `miRDB Score`\n",
    "- `PicTar Score`\n",
    "- `PITA Score`\n",
    "- `TargetScan Score`\n",
    "- `Predicted Sum`\n",
    "- `All Sum`\n",
    "\n",
    "---\n",
    "\n",
    "## ðŸ”¹ Model 3: Post-Clinical Decision Support\n",
    "**Goal:** Support diagnosis after clinical tests, helping in confirmation or doctor-assisted decision making.\n",
    "\n",
    "**Used Features:**\n",
    "- `Endoscopic Images` (as Normal/Abnormal tag or processed features)\n",
    "- `Biopsy Results` (Positive/Negative)\n",
    "- `CT Scan` (Normal/Abnormal)\n",
    "- `helicobacter pylori infection` (Yes/No)\n",
    "\n",
    "> ðŸ’¡ Optionally: Combine outputs from Model 1 & 2 for more context-aware analysis.\n",
    "\n",
    "---\n",
    "\n",
    "## âœ… Target Variable (for all models)\n",
    "- `Label`: Binary (1: Cancer, 0: No Cancer)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "771f3cc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>geographical_location</th>\n",
       "      <th>family_history</th>\n",
       "      <th>smoking_habits</th>\n",
       "      <th>alcohol_consumption</th>\n",
       "      <th>helicobacter_pylori_infection</th>\n",
       "      <th>dietary_habits</th>\n",
       "      <th>existing_conditions</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>Other</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Low_Salt</td>\n",
       "      <td>Chronic Gastritis</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>86</td>\n",
       "      <td>Female</td>\n",
       "      <td>Ethnicity_B</td>\n",
       "      <td>California</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Diabetes</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>68</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>California</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57</td>\n",
       "      <td>Female</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>Other</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Chronic Gastritis</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33</td>\n",
       "      <td>Male</td>\n",
       "      <td>Ethnicity_A</td>\n",
       "      <td>California</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>High_Salt</td>\n",
       "      <td>Diabetes</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age  gender    ethnicity geographical_location  family_history  \\\n",
       "0   43    Male  Ethnicity_A                 Other               1   \n",
       "1   86  Female  Ethnicity_B            California               1   \n",
       "2   68    Male  Ethnicity_A            California               0   \n",
       "3   57  Female  Ethnicity_A                 Other               0   \n",
       "4   33    Male  Ethnicity_A            California               0   \n",
       "\n",
       "   smoking_habits  alcohol_consumption  helicobacter_pylori_infection  \\\n",
       "0               0                    0                              0   \n",
       "1               0                    0                              1   \n",
       "2               1                    1                              0   \n",
       "3               0                    0                              1   \n",
       "4               1                    1                              0   \n",
       "\n",
       "  dietary_habits existing_conditions  label  \n",
       "0       Low_Salt   Chronic Gastritis      0  \n",
       "1      High_Salt            Diabetes      0  \n",
       "2      High_Salt                 NaN      0  \n",
       "3      High_Salt   Chronic Gastritis      0  \n",
       "4      High_Salt            Diabetes      0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "early_risk_columns = [\n",
    "    'age', \n",
    "    'gender', \n",
    "    'ethnicity', \n",
    "    'geographical_location', \n",
    "    'family_history',\n",
    "    'smoking_habits', \n",
    "    'alcohol_consumption',\n",
    "    'helicobacter_pylori_infection', \n",
    "    'dietary_habits',\n",
    "    'existing_conditions',\n",
    "    'label'\n",
    "]\n",
    "\n",
    "early_risk_data = gc_data[early_risk_columns]\n",
    "early_risk_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "21ee7669",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(212354, 11)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "early_risk_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "151cbc30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mature_mirna_acc</th>\n",
       "      <th>mature_mirna_id</th>\n",
       "      <th>target_symbol</th>\n",
       "      <th>target_entrez</th>\n",
       "      <th>target_ensembl</th>\n",
       "      <th>diana_microt</th>\n",
       "      <th>elmmo</th>\n",
       "      <th>microcosm</th>\n",
       "      <th>miranda</th>\n",
       "      <th>mirdb</th>\n",
       "      <th>pictar</th>\n",
       "      <th>pita</th>\n",
       "      <th>targetscan</th>\n",
       "      <th>predicted.sum</th>\n",
       "      <th>all.sum</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>5034</td>\n",
       "      <td>1946626</td>\n",
       "      <td>0.3813</td>\n",
       "      <td>0.1870</td>\n",
       "      <td>0.7864</td>\n",
       "      <td>0.2048</td>\n",
       "      <td>0.5619</td>\n",
       "      <td>0.4382</td>\n",
       "      <td>0.2836</td>\n",
       "      <td>0.9282</td>\n",
       "      <td>4.3243</td>\n",
       "      <td>7.6668</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR234_2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>6901</td>\n",
       "      <td>1503178</td>\n",
       "      <td>0.9586</td>\n",
       "      <td>0.4933</td>\n",
       "      <td>0.9640</td>\n",
       "      <td>0.4980</td>\n",
       "      <td>0.9856</td>\n",
       "      <td>0.1446</td>\n",
       "      <td>0.3754</td>\n",
       "      <td>0.1036</td>\n",
       "      <td>7.9677</td>\n",
       "      <td>1.4833</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR345</td>\n",
       "      <td>MIR345_3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>4014</td>\n",
       "      <td>1787909</td>\n",
       "      <td>0.2695</td>\n",
       "      <td>0.5736</td>\n",
       "      <td>0.6669</td>\n",
       "      <td>0.5404</td>\n",
       "      <td>0.9059</td>\n",
       "      <td>0.8273</td>\n",
       "      <td>0.3509</td>\n",
       "      <td>0.1669</td>\n",
       "      <td>3.7487</td>\n",
       "      <td>3.0468</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR123</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>7351</td>\n",
       "      <td>1310766</td>\n",
       "      <td>0.3723</td>\n",
       "      <td>0.2614</td>\n",
       "      <td>0.9495</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.4299</td>\n",
       "      <td>0.9352</td>\n",
       "      <td>0.7947</td>\n",
       "      <td>0.8670</td>\n",
       "      <td>5.4783</td>\n",
       "      <td>8.8113</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR345</td>\n",
       "      <td>MIR123_1</td>\n",
       "      <td>CDH1</td>\n",
       "      <td>7982</td>\n",
       "      <td>1277058</td>\n",
       "      <td>0.9747</td>\n",
       "      <td>0.7545</td>\n",
       "      <td>0.2632</td>\n",
       "      <td>0.8768</td>\n",
       "      <td>0.6508</td>\n",
       "      <td>0.3377</td>\n",
       "      <td>0.4275</td>\n",
       "      <td>0.9158</td>\n",
       "      <td>1.8092</td>\n",
       "      <td>0.3946</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mature_mirna_acc mature_mirna_id target_symbol  target_entrez  \\\n",
       "0           MIR123        MIR123_1          TP53           5034   \n",
       "1           MIR123        MIR234_2          TP53           6901   \n",
       "2           MIR345        MIR345_3          KRAS           4014   \n",
       "3           MIR123        MIR123_1          KRAS           7351   \n",
       "4           MIR345        MIR123_1          CDH1           7982   \n",
       "\n",
       "   target_ensembl  diana_microt   elmmo  microcosm  miranda   mirdb  pictar  \\\n",
       "0         1946626        0.3813  0.1870     0.7864   0.2048  0.5619  0.4382   \n",
       "1         1503178        0.9586  0.4933     0.9640   0.4980  0.9856  0.1446   \n",
       "2         1787909        0.2695  0.5736     0.6669   0.5404  0.9059  0.8273   \n",
       "3         1310766        0.3723  0.2614     0.9495   0.1342  0.4299  0.9352   \n",
       "4         1277058        0.9747  0.7545     0.2632   0.8768  0.6508  0.3377   \n",
       "\n",
       "     pita  targetscan  predicted.sum  all.sum  label  \n",
       "0  0.2836      0.9282         4.3243   7.6668      0  \n",
       "1  0.3754      0.1036         7.9677   1.4833      0  \n",
       "2  0.3509      0.1669         3.7487   3.0468      0  \n",
       "3  0.7947      0.8670         5.4783   8.8113      0  \n",
       "4  0.4275      0.9158         1.8092   0.3946      0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Genetic_Risk_columns  = [\n",
    "    'mature_mirna_acc', \n",
    "    'mature_mirna_id', \n",
    "    'target_symbol', \n",
    "    'target_entrez',\n",
    "    'target_ensembl', \n",
    "    'diana_microt', \n",
    "    'elmmo', \n",
    "    'microcosm', \n",
    "    'miranda',\n",
    "    'mirdb', \n",
    "    'pictar', \n",
    "    'pita', \n",
    "    'targetscan', \n",
    "    'predicted.sum', \n",
    "    'all.sum',\n",
    "    'label'\n",
    "]\n",
    "\n",
    "genetic_risk_data = gc_data[Genetic_Risk_columns]\n",
    "genetic_risk_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ed768ae7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(212354, 16)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genetic_risk_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "85febdf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>helicobacter_pylori_infection</th>\n",
       "      <th>endoscopic_images</th>\n",
       "      <th>biopsy_results</th>\n",
       "      <th>ct_scan</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Abnormal</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   helicobacter_pylori_infection endoscopic_images biopsy_results   ct_scan  \\\n",
       "0                              0            Normal       Negative  Negative   \n",
       "1                              1            Normal       Negative  Negative   \n",
       "2                              0            Normal       Negative  Negative   \n",
       "3                              1            Normal       Negative  Negative   \n",
       "4                              0          Abnormal       Negative  Negative   \n",
       "\n",
       "   label  \n",
       "0      0  \n",
       "1      0  \n",
       "2      0  \n",
       "3      0  \n",
       "4      0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Post_Clinical_Decision_columns = [\n",
    "    'helicobacter_pylori_infection',\n",
    "    'endoscopic_images', \n",
    "    'biopsy_results', \n",
    "    'ct_scan',\n",
    "    'label'\n",
    "]\n",
    "\n",
    "post_clinical_decision_data = gc_data[Post_Clinical_Decision_columns]\n",
    "post_clinical_decision_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93adc80f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(212354, 5)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "post_clinical_decision_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda61288",
   "metadata": {},
   "source": [
    "## Saving each dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3a6eaa2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All Files saved Successfully âœ…\n"
     ]
    }
   ],
   "source": [
    "early_risk_data.to_csv('../datasets/Early Risk Screening Data/early_screening_data.csv',index = False)\n",
    "\n",
    "genetic_risk_data.to_csv('../datasets/Genetic Risk Model Data/genetic_data.csv',index = False)\n",
    "\n",
    "post_clinical_decision_data.to_csv('../datasets/Post-Clinical Decision Support/clinical_test_data.csv',index = False)\n",
    "\n",
    "print('All Files saved Successfully âœ…')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fc684fd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
